contracts/integrations/nevermined/BalanceTrackerSubscription.sol

```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+import {ERC1155TokenReceiver} from "../../../lib/autonolas-registries/lib/solmate/src/tokens/ERC1155.sol";
+import {IMech} from "../../interfaces/IMech.sol";
+
+interface IERC1155 {
+    /// @dev Gets the amount of tokens owned by a specified account.
+    /// @param account Account address.
+    /// @param tokenId Token Id.
+    /// @return Amount of tokens owned.
+    function balanceOf(address account, uint256 tokenId) external view returns (uint256);
+
+    /// @dev Burns a specified amount of account's tokens.
+    /// @param account Account address.
+    /// @param tokenId Token Id.
+    /// @param amount Amount of tokens.
+    function burn(address account, uint256 tokenId, uint256 amount) external;
+}
+
+/// @dev Only `manager` has a privilege, but the `sender` was provided.
+/// @param sender Sender address.
+/// @param manager Required sender address as a manager.
+error ManagerOnly(address sender, address manager);
+
+/// @dev Provided zero address.
+error ZeroAddress();
+
+/// @dev Provided zero value.
+error ZeroValue();
+
+/// @dev Not enough balance to cover costs.
+/// @param current Current balance.
+/// @param required Required balance.
+error InsufficientBalance(uint256 current, uint256 required);
+
+/// @dev Caught reentrancy violation.
+error ReentrancyGuard();
+
+/// @dev Account is unauthorized.
+/// @param account Account address.
+error UnauthorizedAccount(address account);
+
+/// @dev No incoming msg.value is allowed.
+/// @param amount Value amount.
+error NoDepositAllowed(uint256 amount);
+
+contract BalanceTrackerSubscription is ERC1155TokenReceiver {
+    event MechPaymentCalculated(address indexed mech, uint256 indexed requestId, uint256 deliveryRate, uint256 rateDiff);
+    event CreditsAccounted(address indexed account, uint256 amount);
+
+    // Mech marketplace address
+    address public immutable mechMarketplace;
+    // Subscription NFT
+    address public immutable subscriptionNFT;
+    // Subscription token Id
+    uint256 public immutable subscriptionTokenId;
+
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    // Map of requester => current credit balance
+    mapping(address => uint256) public mapRequesterBalances;
```

Comment:
Note in order to calculate precise credits for requesters - we record credits, not debits, as credits are already bought in subscription.